third_party/amd/lib/TritonAMDGPUToLLVM/TargetInfo.cpp

```diff
@@ -103,34 +105,183 @@ Value TargetInfo::loadDShared(RewriterBase &rewriter, Location loc, Value ptr,
+    return false;
+  }
+
+  Operation *reduxOp = getSingleCombinerFromReduceOp(op);
+  if (!reduxOp)
+    return false;
+
+  auto createDppReduxOp = [&](Type valType, Value &src, int dppCtrl,
+                              int rowMask, int bankMask,
+                              bool boundCtrl) -> Value {
+    // DPP has limited support for data types, so here we need to
+    // cast non-integer types or integer types shorter than 32 bits
+    // to int32, except for fp32.
+    Type actualType = valType;
+    if (!valType.isF32()) {
+      actualType = castToInt(rewriter, loc, src, valType, 32);
+    }
+
+    Value dppResult =
+        rewriter
+            .create<ROCDL::DPPUpdateOp>(loc, actualType, src, src,
+                                        rewriter.getI32IntegerAttr(dppCtrl),
+                                        rewriter.getI32IntegerAttr(rowMask),
+                                        rewriter.getI32IntegerAttr(bankMask),
+                                        rewriter.getBoolAttr(boundCtrl))
+            .getRes();
+
+    if (!valType.isF32()) {
+      castFromInt(rewriter, loc, src, valType, 32);
+      castFromInt(rewriter, loc, dppResult, valType, 32);
+    }
+
+    IRMapping mapping;
+    mapping.map(reduxOp->getOperand(0), src);
+    mapping.map(reduxOp->getOperand(1), dppResult);
+    return rewriter.clone(*reduxOp, mapping)->getResult(0);
+  };
+
+  for (int i = 0; i < acc.size(); i++) {
+    Value buf;
+    auto valType = acc[i].getType();
+
+    /*
+      Here's the implementation of full-wavefront reduction using dpp.
+      https://gpuopen.com/learn/amd-gcn-assembly-cross-lane-operations/
+
+      Each step has a v_mov_dpp instruction following the redux op. In
+      some cases, the lower-level compiler could merge them into single
+      instruction. For example, v_mov_dpp + max => v_max_dpp.
+
+      In DPP, each row consists of 16 consecutive lanes.
+      So the modifier row_shr and row_bcast mean they have the same operations
+      in each row, so in the following instructions, we only take row 0
+      as an example:
+
+      Step 1: Right shift for 8 lanes.
+          lane 8-15 = redux(lane 0-7, lane 8-15)
+
+      Step 2: Right shift for 4 lanes.
+          lane 12-15 = redux(lane 8-11, lane 12-15)
+
+      Step 3: Right shift for 2 lanes.
+          lane 14-15 = redux(lane 12-13, lane 14-15)
+
+      Step 4: Right shift for 1 lane.
+          lane 15 = redux(lane 14, lane 15)
+
+      Step 5: Broadcast lane 15 of each row to all the lanes of its next row.
+          lane 16-31 = redux(lane 15, lane 16-31)
+
+      Step 6: Broadcast lane 31 to lane 32-63.
+          lane 32-63 = redux(lane 31, lane 32-63)
+
+      Now the reduction result is stored in lane 63.
+
+      Step 7: Read the reduction result from lane 63 and broadcast with
+      readlane.
+    */
+
+    // row_shr:8
+    buf = createDppReduxOp(valType, acc[i], 8 + DppCtrl::ROW_SHR0, 0xf, 0xf,
```

Comment:
I'd suggest to create a new more const local variable for these magic numbers given they are basically the same accross all the calls:

```c++
const int allRos = 0xf;
const int allBanks = 0xf;
const bool boundCtrl = true;
```